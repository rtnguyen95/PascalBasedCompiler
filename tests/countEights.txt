! Practice Example !

int countEights(int num1$)
{
    if (num1$ < 8 ) return 0
    endif
    
    int digit = num1$ % 10;   	   ! 18 %10 = 8 !
    num1$ = num1$ / 10;            ! 18 /10 = 1 !
    
    if (digit < 8 or digit > 8)    		   ! 8 == 8     !
    {
        if (num1$ % 10 == 8) 	   ! 1 %10 = 1  !
        
            return 2 + countEights(num1$)
	endif
        
        else 
        
            return 1 + countEights(num1$)
        endelse
    }
    return countEights(num1$);
}

int main()
{
    cout << countEights(8) << endl;
    cout << countEights(818) << endl;
    cout << countEights(8818) << endl;
}





! Practice Example 2 !

int main()
{  
   float rate = 3.0f;
   float initial_balance = 20000.0f;
   double = initial_balance;
   int year = 0;

   while (balance < 2 * initial_balance)  
      balance = balance *  (.25);                ! (1 + rate / 100) !
      year++;
   endwhile

   std::cout << year << std::endl;

   return 0;
}

