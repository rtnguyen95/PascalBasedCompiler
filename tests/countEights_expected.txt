TOKENS        Lexemes

KEYWORD     = int
IDENTIFIER  = countEights
SEPARATOR   = (
KEYWORD     = int
IDENTIFIER  = num1$
SEPARATOR   = )
SEPARATOR   = {
KEYWORD     = if
SEPARATOR   = (
IDENTIFIER  = num1$
OPERATOR    = <
INTEGER     = 8
SEPARATOR   = )
IDENTIFIER  = return
INTEGER     = 0
KEYWORD     = endif
KEYWORD     = int
IDENTIFIER  = digit
OPERATOR    = =
IDENTIFIER  = num1$
OPERATOR    = %
INTEGER     = 10
SEPARATOR   = ;
IDENTIFIER  = num1$
OPERATOR    = =
IDENTIFIER  = num1$
OPERATOR    = /
INTEGER     = 10
SEPARATOR   = ;
KEYWORD     = if
SEPARATOR   = (
IDENTIFIER  = digit
OPERATOR    = <
INTEGER     = 8
KEYWORD     = or
IDENTIFIER  = digit
OPERATOR    = >
INTEGER     = 8
SEPARATOR   = )
SEPARATOR   = {
KEYWORD     = if
SEPARATOR   = (
IDENTIFIER  = num1$
OPERATOR    = %
INTEGER     = 10
OPERATOR    = =
OPERATOR    = =
INTEGER     = 8
SEPARATOR   = )
IDENTIFIER  = return
INTEGER     = 2
OPERATOR    = +
IDENTIFIER  = countEights
SEPARATOR   = (
IDENTIFIER  = num1$
SEPARATOR   = )
KEYWORD     = endif
KEYWORD     = else
IDENTIFIER  = return
INTEGER     = 1
OPERATOR    = +
IDENTIFIER  = countEights
SEPARATOR   = (
IDENTIFIER  = num1$
SEPARATOR   = )
KEYWORD     = endelse
SEPARATOR   = }
IDENTIFIER  = return
IDENTIFIER  = countEights
SEPARATOR   = (
IDENTIFIER  = num1$
SEPARATOR   = )
SEPARATOR   = ;
SEPARATOR   = }
KEYWORD     = int
IDENTIFIER  = main
SEPARATOR   = (
SEPARATOR   = )
SEPARATOR   = {
IDENTIFIER  = cout
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = countEights
SEPARATOR   = (
INTEGER     = 8
SEPARATOR   = )
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = endl
SEPARATOR   = ;
IDENTIFIER  = cout
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = countEights
SEPARATOR   = (
INTEGER     = 818
SEPARATOR   = )
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = endl
SEPARATOR   = ;
IDENTIFIER  = cout
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = countEights
SEPARATOR   = (
INTEGER     = 8818
SEPARATOR   = )
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = endl
SEPARATOR   = ;
SEPARATOR   = }
KEYWORD     = int
IDENTIFIER  = main
SEPARATOR   = (
SEPARATOR   = )
SEPARATOR   = {
KEYWORD     = float
IDENTIFIER  = rate
OPERATOR    = =
FLOAT       = 3.0
IDENTIFIER  = f
SEPARATOR   = ;
KEYWORD     = float
IDENTIFIER  = initial_balance
OPERATOR    = =
FLOAT       = 20000.0
IDENTIFIER  = f
SEPARATOR   = ;
IDENTIFIER  = double
OPERATOR    = =
IDENTIFIER  = initial_balance
SEPARATOR   = ;
KEYWORD     = int
IDENTIFIER  = year
OPERATOR    = =
INTEGER     = 0
SEPARATOR   = ;
KEYWORD     = while
SEPARATOR   = (
IDENTIFIER  = balance
OPERATOR    = <
INTEGER     = 2
OPERATOR    = *
IDENTIFIER  = initial_balance
SEPARATOR   = )
IDENTIFIER  = balance
OPERATOR    = =
IDENTIFIER  = balance
OPERATOR    = *
SEPARATOR   = (
FLOAT       = .25
SEPARATOR   = )
SEPARATOR   = ;
IDENTIFIER  = year
OPERATOR    = +
OPERATOR    = +
SEPARATOR   = ;
IDENTIFIER  = endwhile
IDENTIFIER  = std
SEPARATOR   = :
SEPARATOR   = :
IDENTIFIER  = cout
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = year
OPERATOR    = <
OPERATOR    = <
IDENTIFIER  = std
SEPARATOR   = :
SEPARATOR   = :
IDENTIFIER  = endl
SEPARATOR   = ;
IDENTIFIER  = return
INTEGER     = 0
SEPARATOR   = ;
SEPARATOR   = }
